internal/app/singularity/pull.go

```diff
@@ -54,25 +54,25 @@ func LibraryPull(imgCache *cache.Handle, name, ref, transport, fullURI, libraryU
 	// strip leading "library://" and append default tag, as necessary
 	imageRef := library.NormalizeLibraryRef(fullURI)
 
-	imageName := uri.GetName("library://" + imageRef)
-
 	// check if image exists in library
-	libraryImage, existOk, err := libraryClient.GetImage(context.TODO(), imageRef)
-	if err != nil {
-		return fmt.Errorf("while getting image info: %v", err)
-	}
-	if !existOk {
+	libraryImage, err := libraryClient.GetImage(context.Background(), imageRef)
```

Comment:
hmm...

context.TODO means there's pending work regarding adding proper context handling.

context.Background means this is the topmost context.

They are otherwise identical, the only difference is "documentation".

The key bit in the documentation is this, regarding context.TODO:

> (because the surrounding function has not yet been extended to accept a Context parameter).

What we need to do here is add a signal handler, and in case we receive SIGINT (and other similar signals) we call a cancellation function, so that the download is stopped and any temporary files are removed.

There's a signal handler buried somewhere in this code. That ought to be moved to the main function, where we can create a context.Background, and call the cancellation function as appropriate. This requires extending the CLI run handlers so that they pass this context down. There are other parts of the code that will benefit from this.